metagraph/src/graph/alignment/aligner_labeled.hpp

```diff
@@ -0,0 +1,275 @@
+    target_columns_.reserve(label_counts.size());
+    target_columns_.clear();
+
+    size_t max_count = 0;
+    const auto &label_encoder = anno_graph_.get_annotation().get_label_encoder();
+    for (const auto &[label, count] : label_counts) {
+        max_count = std::max(max_count, count);
+        target_columns_.push_back(label_encoder.encode(label));
+    }
+
+    assert(std::is_sorted(target_columns_.begin(), target_columns_.end()));
+
+    mtg::common::logger->trace("Seed has {} labels", target_columns_.size());
+
+    if (max_count && max_count != path.size()) {
+        mtg::common::logger->warn(
+            "Seed has no common labels. Maximum label multiplicity is {} / {}",
+            max_count, path.size()
+        );
+    }
+}
+
+template <typename NodeType>
+inline auto LabeledColumnExtender<NodeType>
+::fork_extension(NodeType node,
+                 std::function<void(DBGAlignment&&, NodeType)> callback,
+                 score_t min_path_score) -> Edges {
+    if (cached_edge_sets_.count(node))
+        return cached_edge_sets_[node];
+
+    const auto &mat = anno_graph_.get_annotation().get_matrix();
+
+    // get set of outgoing nodes from the parent class
+    auto base_edges = DefaultColumnExtender<NodeType>::fork_extension(
+        node, callback, min_path_score
+    );
+
+    Edges edges;
+
+    // first check the simple case to avoid decoding entire rows
+    if (target_columns_.size() == 1) {
+        for (auto&& edge : base_edges) {
+            if (mat.get(anno_graph_.graph_to_anno_index(edge.first), target_columns_[0]))
+                edges.emplace_back(std::move(edge));
+        }
+
+        cached_edge_sets_[node] = edges;
+
+        return edges;
+    }
+
+    // decode all rows
+    std::vector<AnnotatedDBG::row_index> base_rows;
+    base_rows.reserve(base_edges.size());
+    for (const auto &[out_node, c] : base_edges) {
+        base_rows.push_back(anno_graph_.graph_to_anno_index(out_node));
+    }
+
+    std::vector<Vector<uint64_t>> rows;
+    if (const auto *rd = dynamic_cast<const annot::binmat::IRowDiff*>(&mat)) {
+        rows.resize(base_rows.size());
+        for (size_t i = 0; i < base_rows.size(); ++i) {
+            if (rd->is_anchor(base_rows[i])) {
+                rows[i] = rd->get_diff(base_rows[i]);
+            } else {
+                const auto &dbg_succ = *rd->graph();
+                const auto &boss = dbg_succ.get_boss();
+                if (boss.get_last(dbg_succ.kmer_to_boss_index(base_edges[i].first))) {
+                    // apply diff to target_columns_
+                    auto diff_row = rd->get_diff(anno_graph_.graph_to_anno_index(node));
+                    std::set_difference(target_columns_.begin(), target_columns_.end(),
+                                        diff_row.begin(), diff_row.end(),
+                                        std::back_inserter(rows[i]));
+                } else {
+                    rows[i] = mat.get_row(base_rows[i]);
```

Comment:
always avoid `get_row`. It's inefficient in most of the cases. Always do pre-batching and then call `get_rows`.

Reply:
Doing a batched query here doesn't actually speed things up that much. Each batch would only be size 1-2 on average.